presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergPartitionInsertingPageSource.java

```diff
@@ -0,0 +1,204 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.iceberg;
+
+import com.facebook.presto.common.Page;
+import com.facebook.presto.common.block.Block;
+import com.facebook.presto.common.block.RunLengthEncodedBlock;
+import com.facebook.presto.common.type.Type;
+import com.facebook.presto.hive.HivePartitionKey;
+import com.facebook.presto.spi.ConnectorPageSource;
+import com.facebook.presto.spi.PrestoException;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.function.Function;
+import java.util.stream.IntStream;
+
+import static com.facebook.presto.common.Utils.nativeValueToBlock;
+import static com.facebook.presto.hive.BaseHiveColumnHandle.ColumnType.PARTITION_KEY;
+import static com.facebook.presto.iceberg.IcebergErrorCode.ICEBERG_BAD_DATA;
+import static com.facebook.presto.iceberg.IcebergMetadataColumn.isMetadataColumnId;
+import static com.facebook.presto.iceberg.IcebergUtil.deserializePartitionValue;
+import static com.google.common.base.Throwables.throwIfInstanceOf;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * This class is responsible for filtering the set of columns that should be
+ * read by a page source which performs IO to on-disk files to only those which
+ * are not partition or metadata fields.
+ * <p>
+ * When a new page is requested, this class inserts the partition fields as
+ * RLE-encoded blocks to save memory and IO cost at runtime.
+ */
+public class IcebergPartitionInsertingPageSource
+        implements ConnectorPageSource
+{
+    private final IcebergColumnHandle[] nonPartitionColumnIndexes;
+    private final ConnectorPageSourceWithRowPositions delegateWithPositions;
+    private final ConnectorPageSource delegate;
+    private final Block[] partitionValueBlocks;
+    private final int[] outputIndexes;
+
+    public IcebergPartitionInsertingPageSource(
```

Comment:
nit: for better readability, should we refactor the constructor into a couple of helper methods? Probably one for creating partitionValueBlocks and another one for filtering nonPartitionColumnIndexes and generating outputIndexes.

Reply:
I have refactored it